### ISSUE TITLE
Dek invalidation races

### ISSUE THREAD

**Issue by SamBarker** [kroxylicious/kroxylicious#1649]
the connection close is converted to an error response, but there are several issues that can happen at rollover time:

### 1. the DEK key can be invalidated from the cache multiple times by different threads

As Tom mentioned:
  > We get the dek from the cache, we check it and maybe invalidate the cache entry. The get-and-check pattern is not atomic. So multiple racing threads can end up calling invalidate with the same cache key, but might not be talking about the same cached dek. 

This is addressed in #1643. This means that the freshly created DEK may be removed from the cache immediately. So the exhaustion of 1 DEK could require the loading of several new DEKs before things settle down.

The fix I picked was to add some co-ordination so that if thread A initiates a rotation (DEK destroy for encrypt + invalidate from cache), other threads with a reference to that DEK can chain onto a future to react to the rotation finishing (and it will only be invalidated from the cache once). Retries will take place after invalidation.
 
 ### 2. race condition at destruction time prevents retries and wastes last allocated batch of encryptions
   
Then in fixing that I found further race conditions around destruction:

Thread 1 destroying DEK A for encryption could prevent thread 2 that has successfully been allocated the number of required encryptions from obtaining an Encryptor. Currently the DestroyedDekException thrown isn't handled and would prevent re-attempting, it would just fail immediately
```java
 if (remainingEncryptions.addAndGet(-numEncryptions) >= 0) {
            // some other thread who couldn't get the desired encryptions destroys the DEK before I can get an encryptor
            if (!outstandingCryptors.acquireEncryptorUsage()) {
                throw new DestroyedDekException();
            }
```

So there's two things, the race condition causes wasted encryptions, because a thread can be allocated 100 encryptions and then have the DEK destroyed. Then because we don't handle the exception this doesn't cause a retry.

### 3. retry count is low under test conditions

The retry count of 3 is quite low when you reduce the maximum number of encryptions. Like in my testing at the extreme with 1 encryption per DEK, if you have 5 threads trying to obtain the one DEK then it's easy for a thread to hit the retry limit of 3 as it's only got a 1/5 chance of getting a DEK it's able to use.  Under load with bigger batches of messages we'd probably observe the same thing even with hundreds of encryptions per DEK. With the other race conditions sorted out and a more production sized max encryptions per DEK it shouldn't be an issue. I think we could safely increase it, or make it time based, like a `maxAquireUsableEdekDuration=10s`.

Point 1. is the bigger issue, but 2 and 3 make it hard to test the fix

_Originally posted by @robobario in https://github.com/kroxylicious/kroxylicious/issues/1526#issuecomment-2521909637_

**Comment by tombentley** [comment]
@robobario isn't this fixed by #1643

**Comment by robobario** [comment]
Yes I think this issue was reverse engineered out if the PR once things got a bit messy.
